File: android/src/main/kotlin/com/example/multithread_downloads/HttpsDownloader.kt
package com.example.multithread_downloads

import kotlinx.coroutines.*
import java.io.File
import java.io.FileOutputStream
import kotlin.math.max

class HttpsDownloader {

    suspend fun downloadSingleFile(
        task: DownloadTask,
        onProgress: (Map<String, Any>) -> Unit
    ) {
        val file = File(task.filePath).apply { parentFile?.mkdirs() }
        task.downloadedBytes = if (file.exists()) file.length() else 0L

        task.status = DownloadStatus.DOWNLOADING
        task.startTime = System.currentTimeMillis()
        task.lastSpeedUpdate = task.startTime

        val totalBytes = getFileSize(task)
        task.totalBytes = totalBytes
        sendProgress(task, onProgress)

        repeat(task.retryCount + 1) { attempt ->
            if (task.status != DownloadStatus.DOWNLOADING) return

            try {
                val request = HttpClientConfig.buildRequest(task.url, task.headers, task.downloadedBytes)
                HttpClientConfig.client.newCall(request).execute().use { response ->
                    if (!response.isSuccessful && response.code != 206) {
                        throw Exception("HTTP ${response.code}: ${response.message}")
                    }

                    val body = response.body!!
                    if (task.totalBytes <= 0) {
                        response.header("Content-Length")?.toLongOrNull()?.let {
                            task.totalBytes = it + task.downloadedBytes
                        }
                    }

                    val outputStream = if (task.downloadedBytes > 0) {
                        FileOutputStream(task.filePath, true)
                    } else {
                        FileOutputStream(task.filePath, false)
                    }

                    outputStream.use { output ->
                        val buffer = ByteArray(32768) // Larger buffer for better performance
                        var lastUpdate = 0L
                        var bytesInInterval = 0L

                        body.byteStream().use { input ->
                            var bytesRead: Int
                            while (input.read(buffer).also { bytesRead = it } != -1 &&
                                task.status == DownloadStatus.DOWNLOADING) {

                                output.write(buffer, 0, bytesRead)
                                task.downloadedBytes += bytesRead
                                bytesInInterval += bytesRead

                                val now = System.currentTimeMillis()
                                if (now - lastUpdate > 1000) {
                                    updateSpeedHistory(task, bytesInInterval, now - lastUpdate)
                                    sendProgress(task, onProgress)
                                    lastUpdate = now
                                    bytesInInterval = 0L
                                }
                            }
                        }
                    }

                    if (task.status == DownloadStatus.DOWNLOADING) {
                        task.status = DownloadStatus.COMPLETED
                        sendProgress(task, onProgress)
                    }
                    return
                }
            } catch (e: Exception) {
                if (attempt == task.retryCount) {
                    task.status = DownloadStatus.FAILED
                    task.error = e.message
                    throw e
                }
                delay((500L * (1 shl attempt)) + (0..500).random())
            }
        }
    }

    private suspend fun getFileSize(task: DownloadTask): Long {
        return try {
            val request = HttpClientConfig.buildRequest(task.url, task.headers)
            HttpClientConfig.client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return -1L
                response.header("Content-Length")?.toLongOrNull() ?: -1L
            }
        } catch (e: Exception) {
            -1L
        }
    }

    private fun updateSpeedHistory(task: DownloadTask, bytes: Long, timeMs: Long) {
        val speed = if (timeMs > 0) (bytes * 1000.0 / timeMs) else 0.0
        task.speedHistory.add(speed)
        if (task.speedHistory.size > 10) {
            task.speedHistory.removeAt(0)
        }
    }

    private fun sendProgress(task: DownloadTask, onProgress: (Map<String, Any>) -> Unit) {
        val currentTime = System.currentTimeMillis()
        val timeElapsed = max(1L, currentTime - task.startTime)

        val avgSpeed = if (task.speedHistory.isNotEmpty()) {
            task.speedHistory.average()
        } else {
            (task.downloadedBytes * 1000.0 / timeElapsed)
        }

        val progress = if (task.totalBytes > 0) {
            (task.downloadedBytes * 100.0 / task.totalBytes).toInt()
        } else -1

        onProgress(mapOf(
            "url" to task.url,
            "filePath" to task.filePath,
            "progress" to progress,
            "bytesDownloaded" to task.downloadedBytes,
            "totalBytes" to task.totalBytes,
            "status" to task.status.value,
            "error" to (task.error ?: ""),
            "speed" to avgSpeed
        ))
    }
}

File: ios/Classes/HttpsDownloader.swift
import Foundation

@available(iOS 13.0, *)
class HttpsDownloader {

    func downloadSingleFile(
        task: MTDownloadTask,
        onProgress: @escaping ([String: Any]) -> Void
    ) async throws {
        var task = task

        let fileURL = URL(fileURLWithPath: task.filePath)
        try FileManager.default.createDirectory(at: fileURL.deletingLastPathComponent(), withIntermediateDirectories: true)

        // Resume from existing file if present
        if FileManager.default.fileExists(atPath: task.filePath) {
            let attributes = try FileManager.default.attributesOfItem(atPath: task.filePath)
            task.downloadedBytes = attributes[.size] as? Int64 ?? 0
        } else {
            task.downloadedBytes = 0
        }

        task.status = .downloading
        task.startTime = Date().timeIntervalSince1970 * 1000
        task.lastSpeedUpdate = task.startTime

        let totalBytes = await getFileSize(task: task)
        task.totalBytes = totalBytes
        sendProgress(task: task, onProgress: onProgress)

        // Retry logic
        for attempt in 0...task.retryCount {
            if task.status != .downloading { return }

            do {
                guard let request = HttpClientConfig.buildRequest(
                    url: task.url,
                    headers: task.headers,
                    startByte: task.downloadedBytes
                ) else {
                    throw NSError(domain: "Invalid URL", code: -1)
                }

                let (data, response) = try await HttpClientConfig.client.data(for: request)

                guard let httpResponse = response as? HTTPURLResponse else {
                    throw NSError(domain: "Invalid response", code: -1)
                }

                if !httpResponse.isSuccessful && httpResponse.statusCode != 206 {
                    throw NSError(domain: "HTTP \(httpResponse.statusCode)", code: httpResponse.statusCode)
                }

                // Update total bytes if not known
                if task.totalBytes <= 0 {
                    if let contentLength = httpResponse.value(forHTTPHeaderField: "Content-Length"),
                       let length = Int64(contentLength) {
                        task.totalBytes = length + task.downloadedBytes
                    }
                }

                // Write data to file
                let fileHandle: FileHandle
                if task.downloadedBytes > 0 {
                    // Append mode
                    fileHandle = try FileHandle(forWritingTo: fileURL)
                    fileHandle.seekToEndOfFile()
                } else {
                    // Create new file
                    FileManager.default.createFile(atPath: task.filePath, contents: nil)
                    fileHandle = try FileHandle(forWritingTo: fileURL)
                }

                defer { fileHandle.closeFile() }

                // Process data in chunks with progress updates
                let bufferSize = 32768 // Larger buffer for better performance
                var offset = 0
                var lastUpdate: Double = 0
                var bytesInInterval: Int64 = 0

                while offset < data.count && task.status == .downloading {
                    let chunkSize = min(bufferSize, data.count - offset)
                    let chunk = data.subdata(in: offset..<(offset + chunkSize))

                    fileHandle.write(chunk)
                    task.downloadedBytes += Int64(chunkSize)
                    bytesInInterval += Int64(chunkSize)
                    offset += chunkSize

                    let now = Date().timeIntervalSince1970 * 1000
                    if now - lastUpdate > 1000 {
                        updateSpeedHistory(task: task, bytes: bytesInInterval, timeMs: Int64(now - lastUpdate))
                        sendProgress(task: task, onProgress: onProgress)
                        lastUpdate = now
                        bytesInInterval = 0
                    }
                }

                if task.status == .downloading {
                    task.status = .completed
                    sendProgress(task: task, onProgress: onProgress)
                }
                return

            } catch {
                if attempt == task.retryCount {
                    task.status = .failed
                    task.error = error.localizedDescription
                    throw error
                }

                // Exponential backoff with jitter
                let delay = (500 * (1 << attempt)) + Int.random(in: 0...500)
                try await Task.sleep(nanoseconds: UInt64(delay) * 1_000_000)
            }
        }
    }

    private func getFileSize(task: MTDownloadTask) async -> Int64 {
        do {
            guard let request = HttpClientConfig.buildRequest(url: task.url, headers: task.headers) else {
                return -1
            }

            let (_, response) = try await HttpClientConfig.client.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.isSuccessful else {
                return -1
            }

            if let contentLength = httpResponse.value(forHTTPHeaderField: "Content-Length") {
                return Int64(contentLength) ?? -1
            }

            return -1
        } catch {
            return -1
        }
    }

    private func updateSpeedHistory(task: MTDownloadTask, bytes: Int64, timeMs: Int64) {
        var task = task
        let speed = timeMs > 0 ? (Double(bytes) * 1000.0 / Double(timeMs)) : 0.0
        task.speedHistory.append(speed)
        if task.speedHistory.count > 10 {
            task.speedHistory.removeFirst()
        }
    }

    private func sendProgress(task: MTDownloadTask, onProgress: ([String: Any]) -> Void) {
        let currentTime = Date().timeIntervalSince1970 * 1000
        let timeElapsed = max(1.0, currentTime - task.startTime)

        let avgSpeed: Double
        if !task.speedHistory.isEmpty {
            avgSpeed = task.speedHistory.reduce(0, +) / Double(task.speedHistory.count)
        } else {
            avgSpeed = Double(task.downloadedBytes) * 1000.0 / timeElapsed
        }

        let progress = task.totalBytes > 0 ? Int(Double(task.downloadedBytes) * 100.0 / Double(task.totalBytes)) : -1

        onProgress([
            "url": task.url,
            "filePath": task.filePath,
            "progress": progress,
            "bytesDownloaded": task.downloadedBytes,
            "totalBytes": task.totalBytes,
            "status": task.status.rawValue,
            "error": task.error ?? "",
            "speed": avgSpeed
        ])
    }
}

// Extension for HTTPURLResponse
extension HTTPURLResponse {
    var isSuccessful: Bool {
        return 200...299 ~= statusCode
    }
}
